File: sdk/python/mashgate/webhooks.py
"""Webhook signature verification utility."""

from __future__ import annotations

import hashlib
import hmac
import time

# HookLine replay window: 5 minutes, matching the server (300000 ms).
SIGNATURE_MAX_AGE_MS = 300_000


def verify_webhook_signature(
    payload: str | bytes,
    signature: str,
    secret: str,
    timestamp: str | int | None = None,
    max_age_ms: int = SIGNATURE_MAX_AGE_MS,
) -> bool:
    """Verify a HookLine webhook signature (HMAC-SHA256, timing-safe).

    HookLine signs ``HMAC_SHA256(secret, f"{timestamp}.{body}")`` and sends:

    * ``x-hl-signature``: ``v1=<hex>``
    * ``x-hl-timestamp``: Unix epoch **milliseconds**

    This mirrors the TypeScript/Go SDK verifiers exactly — the signed input is
    ``{timestamp}.{body}`` (NOT the raw body), and the header carries a ``v1=``
    prefix that must be stripped before comparison.

    Args:
        payload: The raw request body.
        signature: The ``x-hl-signature`` header value (``v1=<hex>``).
        secret: The webhook endpoint's signing secret.
        timestamp: The ``x-hl-timestamp`` header value (Unix milliseconds).
            Required — the server signs over ``{timestamp}.{body}``.
        max_age_ms: Reject signatures whose timestamp is older/newer than this
            (replay protection). Pass ``0`` to skip the window check.

    Returns:
        ``True`` iff the ``v1=`` signature matches and, when ``max_age_ms`` is
        set, the timestamp is within the replay window.
    """
    if not signature or not signature.startswith("v1=") or timestamp is None:
        return False

    ts = str(timestamp)

    # Replay window — timestamp is in milliseconds, matching the server.
    if max_age_ms:
        try:
            ts_ms = int(ts)
        except ValueError:
            return False
        now_ms = int(time.time() * 1000)
        if abs(now_ms - ts_ms) > max_age_ms:
            return False

    body = bytes(payload) if isinstance(payload, (bytes, bytearray)) else payload.encode()
    signing_input = ts.encode() + b"." + body

    expected = hmac.new(secret.encode(), signing_input, hashlib.sha256).hexdigest()
    provided = signature[3:]  # strip the "v1=" prefix
    return hmac.compare_digest(expected, provided)

File: sdk/typescript/src/webhooks.ts
/**
 * Verify a HookLine webhook signature using HMAC-SHA256 with timing-safe
 * comparison. Works in both Node.js and browser (WebCrypto API) environments.
 *
 * HookLine signs HMAC-SHA256 over `{timestamp}.{body}` and sends the signature
 * as `x-hl-signature: v1=<hex>` plus an `x-hl-timestamp` (Unix epoch ms). To
 * mitigate replay, a signature whose timestamp is more than `maxAgeMs` from now
 * (in either direction) is rejected; pass `maxAgeMs = 0` to disable the window.
 * This matches the Go and Python SDK verifiers.
 *
 * @param payload   raw request body (string or bytes), NOT re-serialized JSON
 * @param signatureHeader `x-hl-signature` header value (`v1=<hex>`)
 * @param secret    the endpoint signing secret
 * @param timestamp `x-hl-timestamp` header value (Unix epoch milliseconds)
 * @param maxAgeMs  replay window in ms (default 300_000 = 5 min; 0 disables it)
 */
export async function verifyWebhookSignature(
  payload: string | Uint8Array,
  signatureHeader: string,
  secret: string,
  timestamp: string,
  maxAgeMs: number = 300_000,
): Promise<boolean> {
  if (!signatureHeader?.startsWith("v1=") || !timestamp) {
    return false;
  }

  // Replay-window check (fail fast, before the HMAC). The timestamp is Unix
  // epoch milliseconds; reject anything outside ±maxAgeMs of now.
  if (maxAgeMs > 0) {
    const tsMs = Number(timestamp);
    if (!Number.isFinite(tsMs) || Math.abs(Date.now() - tsMs) > maxAgeMs) {
      return false;
    }
  }

  const encoder = new TextEncoder();
  const payloadBytes = typeof payload === "string" ? encoder.encode(payload) : payload;
  const prefixBytes = encoder.encode(`${timestamp}.`);
  const signingInput = new Uint8Array(prefixBytes.length + payloadBytes.length);
  signingInput.set(prefixBytes, 0);
  signingInput.set(payloadBytes, prefixBytes.length);
  const secretBytes = encoder.encode(secret);

  const key = await crypto.subtle.importKey(
    "raw",
    secretBytes,
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );

  const signatureBuffer = await crypto.subtle.sign("HMAC", key, signingInput);
  const computed = bufferToHex(signatureBuffer);
  const provided = signatureHeader.slice(3);

  return timingSafeEqual(computed, provided);
}

function bufferToHex(buffer: ArrayBuffer): string {
  const bytes = new Uint8Array(buffer);
  let hex = "";
  for (let i = 0; i < bytes.length; i++) {
    hex += bytes[i].toString(16).padStart(2, "0");
  }
  return hex;
}

function timingSafeEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;

  let result = 0;
  for (let i = 0; i < a.length; i++) {
    result |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return result === 0;
}
